service/utils/file.py

```diff
@@ -0,0 +1,234 @@
+                                    .encode()
+                                    .decode("unicode_escape")
+                                    .lower()
+                                }
+                            )
+
+                        # json.dumps will output empty values in the JSON (which is technically incorrect) { "a": "" }
+                        # so we force empty values to None so they will appear as null
+                        # in the end JSON output { "a": null }
+                        if not value:
+                            row.update({key: None})
+
+                        # pull out requested csv headers if header is requested and header exists or add all the headers
+                        if (self.config.get("h") and key in self.config.get("h")) or (
+                            not self.config.get("h")
+                        ):
+                            rl[key] = row[key]
+
+                        # if prune is set, remove any columns with a value of None
+                        if self.config.get("prune") and row[key] is None and key in rl:
+                            del rl[key]
+
+                    # if prune is set and we are sorting, we have to ensure the rows have all the columns to sort on
+                    if self.config.get("prune") and self.config.get("s"):
+                        cur_kz = list(rl.keys())
+                        srt_on = list(map(itemgetter("column"), self.config.get("s")))
+                        if all(item in cur_kz for item in srt_on):
+                            dl.append(rl)
+                    else:
+                        dl.append(rl)
+
+                self.total_data = len(dl)
+
+        except TypeError:
+            raise TypeError("Invalid Universe File")
+
+        return self.filter_data(dl)
+
+    def sort_results(self, results, srt_ordr=None):
+        """
+        Custom sorting function that allows for sorting None while also maintaining locale aware sorting
+        config["nulls"] will allow for sorting None and putting them at the front or end of the list
+
+        l = [1, 3, 2, 5, 4, None, 7]
+        print('Last = ', sorted(l, key=lambda x: (x is None, x)))
+        Last = [1, 2, 3, 4, 5, 7, None]
+        print('First = ', sorted(l, key=lambda x: (x is not None, x)))
+        First = [None, 1, 2, 3, 4, 5, 7]
+
+        :param results: (list) e.g. [{'name': 'richard jones (earth-616)', 'appearances': 590, 'year': 1962}]
+        :param srt_ordr: (dict) e.g. {'column': 'name', 'sort': False}
+        :return: (tuple) e.g. (False, True, 1969) tuples are sorted by item type, this means that all non-None elements
+        will come first (since False < True), and then be sorted by value.
+        """
+        srt_dict = srt_ordr if srt_ordr is not None else self.config.get("s")
+
+        for i in reversed(srt_dict):
+            results.sort(key=lambda itm: self.sort_i18n_str(itm, i), reverse=i["sort"])
```

Comment:
If `.sort_i18n_str()` ever fails, it would be nice to debug the sorting. I'm not familiar with a way to debug `lambda`s. I recommend turning this into function. For example:

```python
foo = [1, 4, 1, 3, 4, 4, 1]

def ones_in_the_back(item):
  if item == 1:
    return 1
  else:
    return 0

foo.sort(key=ones_in_the_back)

print(foo)
# >>> [4, 3, 4, 4, 1, 1, 1]
```

Reply:
So I could use some advice here. 

From what I've researched the, generally accepted way to go about solving the issue with the sort and passing two parameters to the method is the way I've done it e.g. `results.sort(key=lambda itm: self.sort_i18n_str(itm, i), reverse=i["sort"])`

The srt_dict is a list of dictionaries that tell the function what to sort on and what direction

```
[
   {
      'column': 'name', 
       'sort': True
   },
   {
      'column': 'eye', 
       'sort': False
   }
]
```
= Name: Desc, Eye, Asc

The first loop is iterating over that list. I have to pass in each dictionary to the sort_i18n_str function to determine which direction the None values will be listed in.

`results.sort(key=self.sort_i18n_str, reverse=i["sort"])` only allows 1 positional argument (which is the method name passing the row I'm sorting on), 

To pass the second argument to the `self.sort_i18n_str` I have to:

1. Update an internal class variable with the correct value and pull from that inside the `sort_i18n_str` function which seems dirty
1. Use functools.partial which looks just like the lambda `results.sort(key=partial(self.sort_i18n_str, row, i), reverse=i["sort"])` 

Any thoughts?

Reply:
> results.sort(key=self.sort_i18n_str, reverse=i["sort"]) only allows 1 positional argument (which is the method name passing the row I'm sorting on)

I don't think I understand this sentence. When a function is passed as a `key`, the function's parameter is the iterable in the loop. Instead of passing in multiple inputs, it would get just 1, and the function would then have to unpack that iterable.

Reply:
```python

results = [
{
        "name": "human torch (android) (earth-616)",
        "sex": "male characters",
        "year": 1939
    },
    {
        "name": "thomas raymond (earth-616)",
        "sex": "male characters",
        "year": 1940
    }
]
d = [
   {
      'column': 'name', 
       'sort': True
   },
   {
      'column': 'eye', 
       'sort': False
   }
]

for direction in reversed(d):
            results.sort(key=lambda itm: self.sort_i18n_str(itm, direction), reverse=direction["sort"])
...

def sort_i18n_str(itm, direction):
  print (itm, direction)

# {"name": "human torch (android) (earth-616)", "sex": "male characters", "year": 1939"}, {"column": "name",   "sort": True
```

The labmda in the above scenario allows me to to pass both the itm ({"name": "human torch (android) (earth-616)", "sex": "male characters", "year": 1939"}) and the direction ({"column": "name",   "sort": True}) to the sort method

the way you proposed it, it can be an either or situation, but not both

```python

for direction in reversed(d):
            results.sort(key=self.sort_i18n_str, reverse=direction["sort"])

def sort_i18n_str(itm):
  print (itm)


# {"name": "human torch (android) (earth-616)", "sex": "male characters", "year": 1939"}
```

OR

```python

for direction in reversed(d):
            results.sort(key=self.sort_i18n_str(direction), reverse=direction["sort"])

def sort_i18n_str(itm):
  print (itm)

# {"column": "name",   "sort": True}
```

key can only have one thing passed to it e.g. a function name (or an anonymous function), and the first position argument in the function is the item being passed to it. However, if you pass a function name with a parameter e.g. self.sort_i18n_str(direction) the first positional argument is no longer the item but the thing you passed in e.g. direction

Reply:
```python
directions = [True, False, True, False]
foo = [1, 4, 1, 3, 4, 4, 1]


for direction in directions:
  def ones_in_the_back(item):
    if item == 1 and direction:
      return 1
    else:
      return 0

  foo.sort(key=ones_in_the_back)

print(foo)
# >>> [4, 3, 4, 4, 1, 1, 1]
```